### DISCUSSION TITLE
ABC's with Final Attributes

### DISCUSSION THREAD

**Discussion by nicwest** [microsoft/pyright#9578] [Q&A]
I'm trying to understand why pyright doesn't like this.

```python
from abc import ABC

class Point:
    x: int

    def __init__(self) -> None:
        self.x = 1

class BasePoint(ABC):
    x: int

class SubPoint(BasePoint):

    def __init__(self) -> None:
        self.x = 1

class ImmutablePoint:
    x: Final[int]

    def __init__(self) -> None:
        self.x = 1

class BaseImmutablePoint(ABC):
    x: Final[int]      # "x" is declared Final, but value is not assigned


class SubImmutablePoint(BaseImmutablePoint):
    def __init__(self) -> None:
        self.x = 1     # "x" cannot be redeclared because parent class "BaseImmutablePoint" declares it as Final
```
I would expect the last two variations to be consistent with the non ABC/final variations

**Comment by erictraut** [comment]
Pyright is following the typing spec in this case. Mypy produces the same results.

Please refer to the [this section of the spec](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#uppercase-final) for information about `Final`.

The `ABC` variant you've defined here is not equivalent to the non-ABC variant. In the non-ABC variant, you've used `Final` in the subclass. In the `ABC` variant, you've used `Final` in the parent class, and you've failed to provide a value for it — and then overridden its value in the subclass. 

Perhaps the draft [PEP 767](https://peps.python.org/pep-0767/) provides the functionality you're looking for here.
